Moving a message away from the `print` to `logging`

adamb0mb opened the pull request:
The print statement ended up in the mapper's output. If you run something with the `inline` runner, it works fine, but if you run something with the `local` runner, it blows up.

I think _any_ `print` statement w/in the mrjob context is going to break things. I can confirm it breaks things if you include extra `print`s in the mapper.

(Including the call stack, so future searchers have something to look for)
> Traceback (most recent call last):
  File "myapp.py", line 60, in <module>
    MyApp.run()
  File "/private/var/folders/w8/l_0frgpx5y18fdjnwl05x73h0000gp/T/myapp.AdamP.20171003.032506.719205/job_local_dir/0/mapper/0/mrjob.zip/mrjob/job.py", line 439, in run
  File "/private/var/folders/w8/l_0frgpx5y18fdjnwl05x73h0000gp/T/myapp.AdamP.20171003.032506.719205/job_local_dir/0/mapper/0/mrjob.zip/mrjob/job.py", line 451, in execute
  File "/private/var/folders/w8/l_0frgpx5y18fdjnwl05x73h0000gp/T/myapp.AdamP.20171003.032506.719205/job_local_dir/0/mapper/0/mrjob.zip/mrjob/job.py", line 612, in run_combiner
  File "/Users/AdamP/git/cc-mrjob/mrcc.py", line 44, in combiner
    for key_val in self.reducer(key, value):
  File "myapp.py", line 54, in reducer
    for pagedata in pagedatas:
  File "/private/var/folders/w8/l_0frgpx5y18fdjnwl05x73h0000gp/T/myapp.AdamP.20171003.032506.719205/job_local_dir/0/mapper/0/mrjob.zip/mrjob/job.py", line 611, in <genexpr>
  File "/private/var/folders/w8/l_0frgpx5y18fdjnwl05x73h0000gp/T/myapp.AdamP.20171003.032506.719205/job_local_dir/0/mapper/0/mrjob.zip/mrjob/job.py", line 708, in read_lines
  File "/private/var/folders/w8/l_0frgpx5y18fdjnwl05x73h0000gp/T/myapp.AdamP.20171003.032506.719205/job_local_dir/0/mapper/0/mrjob.zip/mrjob/protocol.py", line 90, in read
ValueError: need more than 1 value to unpack
Step 1 of 1 failed: Command '['sh', '-ex', 'setup-wrapper.sh', '/Users/AdamP/git/cc-mrjob/env/bin/python', 'myapp.py', '--step-num=0', '--combiner', '-r', 'local']' returned non-zero exit status 1

sebastian-nagel commented:
Thanks @adamb0mb, definitely the print statement isn't the right way for logging:
- if running as streaming job (all [runners](https://pythonhosted.org/mrjob/guides/runners.html) except the [InlineMRJobRunner](https://pythonhosted.org/mrjob/runners-inline.html#mrjob.inline.InlineMRJobRunner) everything written to stdout is passed as stream to the output or to reducers and combiners. These fail if their input isn't a key-value pair (most log message aren't).
- I also miss proper timestamps which enable debugging of performance issues

However, my experience with the logging module and mrjob on a Hadoop/EMR cluster: you have to configure the logger so that it writes to stderr (cf. [this post](https://stackoverflow.com/questions/26123152/mrjob-setup-logging-on-emr) or [that](https://groups.google.com/d/msg/mrjob/0VL02lfWTv0/39oZ-ybQIxwJ)). I've just decided to write a [trivial log function](https://github.com/commoncrawl/webarchive-indexing/blob/boto3/indexwarcsjob.py#L96) for that purpose. But the right way is probably to use logging in combination with `log_to_stream` of the [MRJob utility functions](https://pythonhosted.org/mrjob/utils-util.html). Can you try it? I'll then verify whether it works on a Hadoop cluster.

adamb0mb commented:
I spent 30 minutes trying to get the `log_to_stream` hooked up, but couldn't figure out how I was supposed to be using it. I'll spend some more time tonight and see what I can do.

adamb0mb commented:
OK. 2 things.

First, I've set up logging "the right way," for `MRJob` as far as I can tell. But, because of a [known issue in MRJob](https://github.com/Yelp/mrjob/issues/1551), it causes duplicate log messages. I'm going to leave the dupes in for now, because I assume the referenced issue will be closed at some point. Also, I added some formatting to the log output, to get some timestampping on it.

Second, I got tired of `pylint` yelling at me, so I cleaned a couple of things up (and added some comments). They're pretty basic but they may be more than you're looking for. I'm not attached to 'em so if you're not into it, I can just strip 'em out.

adamb0mb commented:
I fixed the duplicate logging.

sebastian-nagel commented:
Thanks @adamb0mb for digging into the problem of duplicate log messages and the doc strings!

I'll verified the solution locally but I want to test on a cluster. I'll do this next week or the week after. Thanks!

adamb0mb commented:
I'm planning on running on an EMR today or tomorrow (I've never run an EMR job before, so I'll let you know if I'm successful :-))

sebastian-nagel commented:
Just in case you're having issues accessing data on S3: #18 - another user of cc-mrjob recently reported issues and I've also seen 99% of tasks failing.

adamb0mb commented:
Jeez! Thanks for the heads up!

sebastian-nagel commented:
Hi @adamb0mb,
I've verified that your solution works on Hadoop (Cloudera CDH 5.13.0). I've also changed the logging for the used derived class SitemapExtractor.
After the first run I've decided
- to adapt the logging format to the default Hadoop log format (e.g., [YARN log4j.properties](https://github.com/apache/hadoop/blob/trunk/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/resources/container-log4j.properties)): use space as separator and add the class name. That simplifies grepping over all logs and makes clear whether a message comes from CCJob or a derived class.
- set the logger name explicitly to the class name because it would be just "__main__" when using the local runner . With other runners (hadoop, emr) `name=__name__` works, but it's confusing when the test environment does behave as expected.

Logs look know like:
```
Container: container_1507808170811_0205_01_000226 on ip-10-136-183-252.ec2.internal_37199
===========================================================================================
LogType:stderr
...
2017-10-26 17:02:59,983 INFO CCJob: Loading s3://commoncrawl/crawl-data/CC-MAIN-2017-43/segments/1508187823630.63/robotstxt/CC-MAIN-20171020025810-20171020045810-00366.warc.gz
2017-10-26 17:03:03,034 WARNING SitemapExtractor: Unexpectedly large number of sitemaps (19540) announced in robots.txt URL: http://ied8.fbftt.com/robots.txt
...

LogType:syslog
...
2017-10-26 17:02:58,200 INFO [main] org.apache.hadoop.streaming.PipeMapRed: PipeMapRed exec [/bin/sh, -ex, setup-wrapper.sh, python, sitemaps_from_robotstxt.py, --step-num=0, --mapper, -r, hadoop]
...
```

I hope you're fine with this changes, see [diff](https://github.com/commoncrawl/cc-mrjob/compare/master...commoncrawl:20-move-to-logging). I'm going to merge it soon into master. Thanks for your work!

adamb0mb commented:
Looks awesome... Thanks!

--
Adam

On Thu, Oct 26, 2017 at 1:19 PM, Sebastian Nagel <notifications@github.com>
wrote:

> Hi @adamb0mb <https://github.com/adamb0mb>,
> I've verified that your solution works on Hadoop (Cloudera CDH 5.13.0).
> I've also changed the logging for the used derived class SitemapExtractor.
> After the first run I've decided
>
>    - to adapt the logging format to the default Hadoop log format (e.g., YARN
>    log4j.properties
>    <https://github.com/apache/hadoop/blob/trunk/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/resources/container-log4j.properties>):
>    use space as separator and add the class name. That simplifies grepping
>    over all logs and makes clear whether a message comes from CCJob or a
>    derived class.
>    - set the logger name explicitly to the class name because it would be
>    just "*main*" when using the local runner . With other runners
>    (hadoop, emr) name=__name__ works, but it's confusing when the test
>    environment does behave as expected.
>
> Logs look know like:
>
> Container: container_1507808170811_0205_01_000226 on ip-10-136-183-252.ec2.internal_37199
> ===========================================================================================
> LogType:stderr
> ...
> 2017-10-26 17:02:59,983 INFO CCJob: Loading s3://commoncrawl/crawl-data/CC-MAIN-2017-43/segments/1508187823630.63/robotstxt/CC-MAIN-20171020025810-20171020045810-00366.warc.gz
> 2017-10-26 17:03:03,034 WARNING SitemapExtractor: Unexpectedly large number of sitemaps (19540) announced in robots.txt URL: http://ied8.fbftt.com/robots.txt
> ...
>
> LogType:syslog
> ...
> 2017-10-26 17:02:58,200 INFO [main] org.apache.hadoop.streaming.PipeMapRed: PipeMapRed exec [/bin/sh, -ex, setup-wrapper.sh, python, sitemaps_from_robotstxt.py, --step-num=0, --mapper, -r, hadoop]
> ...
>
> I hope you're fine with this changes, see diff
> <https://github.com/commoncrawl/cc-mrjob/compare/master...commoncrawl:20-move-to-logging>.
> I'm going to merge it soon into master. Thanks for your work!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/commoncrawl/cc-mrjob/pull/20#issuecomment-339788680>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWuC-36fGEP6ixt0hIeut0EdTWa_tjiks5swOk9gaJpZM4PrlKa>
> .
>
